File: routes/categories.js
const express = require('express');
const router = express.Router();
const db = require('../database');

router.get('/categories', async (req, res) => {
  try {
    const result = await db.query('SELECT name FROM categories ORDER BY name');

    let categories = result.rows.map(row => row.name);

    if (categories.length === 0) {
      const defaultCategories = [
        'Food & Drink',
        'Shopping',
        'Travel',
        'Health & Wellness',
        'Groceries',
        'Bills & Utilities',
        'Entertainment',
        'Personal',
        'Professional Services',
        'Uncategorized'
      ];
      categories = defaultCategories;
    }

    res.json({ categories });
  } catch (error) {
    console.error('Error reading categories:', error);
    res.status(500).json({ error: 'Failed to read categories' });
  }
});

router.post('/categories', async (req, res) => {
  try {
    const { categories } = req.body;

    const client = await db.beginTransaction();
    try {
      await client.query('DELETE FROM categories');

      for (const category of categories) {
        await client.query(
          'INSERT INTO categories (name) VALUES ($1) ON CONFLICT (name) DO NOTHING',
          [category]
        );
      }

      await db.commitTransaction(client);
    } catch (error) {
      await db.rollbackTransaction(client);
      throw error;
    }

    res.json({ success: true, count: categories.length });
  } catch (error) {
    console.error('Error saving categories:', error);
    res.status(500).json({ error: 'Failed to save categories' });
  }
});

module.exports = router;

File: backend/app/routes/categories.py
from fastapi import APIRouter, Depends
from sqlalchemy import delete, select, text
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies.auth import get_current_household_id
from app.models.category import Category
from app.schemas.category import CategoriesSaveRequest

router = APIRouter(prefix="/api", tags=["categories"])

DEFAULT_CATEGORIES = [
    "Food & Drink",
    "Shopping",
    "Travel",
    "Health & Wellness",
    "Groceries",
    "Bills & Utilities",
    "Entertainment",
    "Personal",
    "Professional Services",
    "Uncategorized",
]


@router.get("/categories")
async def get_categories(
    household_id: str = Depends(get_current_household_id),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Category.name)
        .where(Category.household_id == household_id)
        .order_by(Category.name)
    )
    categories = [row[0] for row in result.all()]
    if not categories:
        categories = DEFAULT_CATEGORIES
    return {"categories": categories}


@router.post("/categories")
async def save_categories(
    body: CategoriesSaveRequest,
    household_id: str = Depends(get_current_household_id),
    db: AsyncSession = Depends(get_db),
):
    await db.execute(delete(Category).where(Category.household_id == household_id))
    for name in body.categories:
        # id is just the name (matches the migration's surrogate-id backfill)
        await db.execute(
            text(
                "INSERT INTO categories (id, name, household_id) "
                "VALUES (:id, :name, :hid) "
                "ON CONFLICT (household_id, name) DO NOTHING"
            ),
            {"id": name, "name": name, "hid": household_id},
        )
    await db.commit()
    return {"success": True, "count": len(body.categories)}


@router.get("/labels")
async def get_labels(
    household_id: str = Depends(get_current_household_id),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        text("""
            SELECT DISTINCT lbl
            FROM transactions,
                jsonb_array_elements_text(
                    COALESCE(labels, '[]'::jsonb)
                ) AS lbl
            WHERE transactions.household_id = :hid
            ORDER BY lbl
        """),
        {"hid": household_id},
    )
    labels = [row[0] for row in result.all()]
    return {"labels": labels}
